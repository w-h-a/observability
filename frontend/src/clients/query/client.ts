export interface IClient {
	get<T = any>(url: string): Promise<{ data: T }>;
}
