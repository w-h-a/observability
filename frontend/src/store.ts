import { applyMiddleware, combineReducers, compose, createStore } from "redux";
import { thunk } from "redux-thunk";

export const reducers = combineReducers({});

export const store = createStore(reducers, compose(applyMiddleware(thunk)));
