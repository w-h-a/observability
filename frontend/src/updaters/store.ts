import { combineReducers, configureStore } from "@reduxjs/toolkit";
import { ServicesUpdater } from "./services/servicesTable";
import { TimeUpdater } from "./time/time";
import { ServiceUpdater } from "./service/service";

const reducer = combineReducers({
	maxMinTime: TimeUpdater.MaxMinTimeReducer,
	services: ServicesUpdater.ServicesReducer,
	endpoints: ServiceUpdater.EndpointsReducer,
});

export const store = configureStore({
	reducer: reducer,
});

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;
