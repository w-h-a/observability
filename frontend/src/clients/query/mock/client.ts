import { IClient } from "../client";

export class Client implements IClient {
	async get<T = any>(url: string): Promise<{ data: T }> {
		return {
			data: [] as T,
		};
	}
}
