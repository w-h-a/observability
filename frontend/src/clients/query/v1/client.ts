import axios, { AxiosInstance } from "axios";
import { IClient } from "../client";

export class Client implements IClient {
	private client: AxiosInstance = axios.create();

	async get<T = any>(url: string): Promise<{ data: T }> {
		const result = await this.client.get<T>(url);
		return result;
	}
}
