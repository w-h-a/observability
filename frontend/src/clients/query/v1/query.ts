import { Config } from "../../../config/config";
import { IClient } from "../client";

export class Query {
	static async GetServices<T = any>(
		client: IClient,
		start: number,
		end: number,
	): Promise<{ data: T }> {
		try {
			const path = `/services`;

			const query = `?start=${start}&end=${end}`;

			return await client.get<T>(
				`${Config.GetInstance().get("baseUrl")}${path}${query}`,
			);
		} catch (err: unknown) {
			console.log(`query client failed to retrieve services: ${err}`);
			throw err;
		}
	}

	static async GetEndpoints<T = any>(
		client: IClient,
		start: number,
		end: number,
		serviceName: string,
	): Promise<{ data: T }> {
		try {
			const path = `/service/endpoints`;

			const query = `?start=${start}&end=${end}&service=${serviceName}`;

			return await client.get<T>(
				`${Config.GetInstance().get("baseUrl")}${path}${query}`,
			);
		} catch (err: unknown) {
			console.log(`query client failed to retrieve endpoints: ${err}`);
			throw err;
		}
	}
}
