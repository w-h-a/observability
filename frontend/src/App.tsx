import { createContext, useContext } from "react";
import { Provider } from "react-redux";
import { BrowserRouter, Redirect, Route, Switch } from "react-router-dom";
import { Layout } from "antd";
import { Content, Footer } from "antd/es/layout/layout";
import { SideNav } from "./views/Nav/SideNav";
import { TopNav } from "./views/Nav/TopNav";
import { ServicesTable } from "./views/Services/ServicesTable";
import { Client } from "./clients/query/v1/client";
import { IClient } from "./clients/query/client";
import { store } from "./updaters/store";

export const ClientContext = createContext<{ queryClient: IClient }>({
	queryClient: new Client(),
});

export const App = () => {
	return (
		<Provider store={store}>
			<BrowserRouter basename="/">
				<Layout style={{ minHeight: "100vh" }}>
					<SideNav />
					<Layout className="site-layout">
						<Content style={{ margin: "0 16px" }}>
							<TopNav />
							<ClientContext.Provider value={useContext(ClientContext)}>
								<Switch>
									<Route path="/application" component={ServicesTable} />
									<Route
										path="/"
										exact
										render={() => {
											return <Redirect to="/application" />;
										}}
									/>
								</Switch>
							</ClientContext.Provider>
						</Content>
						<Footer>Trace-Blame 2025</Footer>
					</Layout>
				</Layout>
			</BrowserRouter>
		</Provider>
	);
};
