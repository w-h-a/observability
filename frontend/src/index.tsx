import React from "react";
import ReactDOM from "react-dom/client";
import App from "./App";
import { Provider } from "react-redux";
import { store } from "./store";
import { BrowserRouter } from "react-router-dom";

const root = ReactDOM.createRoot(
	document.getElementById("root") as HTMLElement,
);

root.render(
	<Provider store={store}>
		<React.StrictMode>
			<BrowserRouter basename="/">
				<App />
			</BrowserRouter>
		</React.StrictMode>
	</Provider>,
);
