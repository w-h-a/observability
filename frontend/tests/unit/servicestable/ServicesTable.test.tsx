import { Provider } from "react-redux";
import renderer, { create, act } from "react-test-renderer";
import { IClient } from "../../../src/clients/query/client";
import { Client } from "../../../src/clients/query/mock/client";
import { store } from "../../../src/updaters/store";
import { ServicesTable } from "../../../src/views/Services/ServicesTable";
import { TestCase } from "../testcase";
import { ClientContext } from "../../../src/App";
import { BrowserRouter } from "react-router-dom";

vi.mock("../../../src/updaters/time/utils", () => {
	return {
		Now: vi.fn(() => 1736393070 * 1000),
	};
});

describe("ServicesTable", () => {
	const mockQueryClient: IClient = new Client();

	const testCases: TestCase[] = [
		{
			When: "when: the user goes to the services table and there are services",
			GetStub: function <T>(url: string): Promise<{ data: T }> {
				return new Promise((resolve) => {
					resolve({
						data: [
							{
								serviceName: "testService1",
								p99: 4000 * 1000000,
								callRate: 0.5,
								errorRate: 0.0,
							},
							{
								serviceName: "testService2",
								p99: 2000 * 1000000,
								callRate: 0.75,
								errorRate: 0.01,
							},
						] as T,
					});
				});
			},
			Element: <ServicesTable />,
			Then: "then: we render a table with the services",
			ClientCalledTimes: 1,
			ClientCalledWith: [
				"http://localhost:4000/api/v1/services?start=1736392170&end=1736393070",
			],
		},
		{
			When:
				"when: the user goes to the services table and the query client fails for some reason",
			GetStub: function <T>(url: string): Promise<{ data: T }> {
				return new Promise((_, reject) => {
					reject(new Error("whoops!"));
				});
			},
			Element: <ServicesTable />,
			Then: "then: we render the error table",
			ClientCalledTimes: 1,
			ClientCalledWith: [
				"http://localhost:4000/api/v1/services?start=1736392170&end=1736393070",
			],
		},
	];

	testCases.forEach((testCase) => {
		describe(testCase.When, () => {
			let view: renderer.ReactTestRenderer;

			beforeEach(async () => {
				mockQueryClient.get = vi.fn().mockImplementation(testCase.GetStub);

				await act(async () => {
					view = create(
						<Provider store={store}>
							<BrowserRouter>
								<ClientContext.Provider value={{ queryClient: mockQueryClient }}>
									{testCase.Element}
								</ClientContext.Provider>
							</BrowserRouter>
						</Provider>,
					);
				});
			});

			afterEach(() => {
				vi.clearAllMocks();
			});

			test(testCase.Then, async () => {
				expect(mockQueryClient.get).toHaveBeenCalledTimes(
					testCase.ClientCalledTimes,
				);

				for (let count = 1; count <= testCase.ClientCalledTimes; count++) {
					expect(mockQueryClient.get).toHaveBeenCalledWith(
						testCase.ClientCalledWith[count - 1],
					);
				}

				expect(view.toJSON()).toMatchSnapshot();
			});
		});
	});
});
