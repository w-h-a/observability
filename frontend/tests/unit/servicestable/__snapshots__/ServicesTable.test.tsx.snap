// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ServicesTable > when: the user goes to the services table and the query client fails for some reason > then: we render the error table 1`] = `
<div>
  <div
    class="ant-table-wrapper css-dev-only-do-not-override-7ny38l"
  >
    <div
      class="ant-spin-nested-loading css-dev-only-do-not-override-7ny38l"
    >
      <div
        class="ant-spin-container"
      >
        <div
          class="ant-table css-dev-only-do-not-override-7ny38l"
        >
          <div
            class="ant-table-container"
          >
            <div
              class="ant-table-content"
            >
              <table
                style="table-layout: auto;"
              >
                <colgroup />
                <thead
                  class="ant-table-thead"
                >
                  <tr>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Service
                    </th>
                    <th
                      aria-label="P99 latency (in ms)"
                      class="ant-table-cell ant-table-column-has-sorters"
                      scope="col"
                      tabindex="0"
                    >
                      <div
                        class="ant-table-column-sorters"
                      >
                        <span
                          class="ant-table-column-title"
                        >
                          P99 latency (in ms)
                        </span>
                        <span
                          class="ant-table-column-sorter ant-table-column-sorter-full"
                        >
                          <span
                            aria-hidden="true"
                            class="ant-table-column-sorter-inner"
                          >
                            <span
                              aria-label="caret-up"
                              class="anticon anticon-caret-up ant-table-column-sorter-up"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="caret-up"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                />
                              </svg>
                            </span>
                            <span
                              aria-label="caret-down"
                              class="anticon anticon-caret-down ant-table-column-sorter-down"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="caret-down"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                />
                              </svg>
                            </span>
                          </span>
                        </span>
                      </div>
                    </th>
                    <th
                      aria-label="Error Rate (in %)"
                      class="ant-table-cell ant-table-column-has-sorters"
                      scope="col"
                      tabindex="0"
                    >
                      <div
                        class="ant-table-column-sorters"
                      >
                        <span
                          class="ant-table-column-title"
                        >
                          Error Rate (in %)
                        </span>
                        <span
                          class="ant-table-column-sorter ant-table-column-sorter-full"
                        >
                          <span
                            aria-hidden="true"
                            class="ant-table-column-sorter-inner"
                          >
                            <span
                              aria-label="caret-up"
                              class="anticon anticon-caret-up ant-table-column-sorter-up"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="caret-up"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                />
                              </svg>
                            </span>
                            <span
                              aria-label="caret-down"
                              class="anticon anticon-caret-down ant-table-column-sorter-down"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="caret-down"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                />
                              </svg>
                            </span>
                          </span>
                        </span>
                      </div>
                    </th>
                    <th
                      aria-label="Requests Per Second"
                      class="ant-table-cell ant-table-column-has-sorters"
                      scope="col"
                      tabindex="0"
                    >
                      <div
                        class="ant-table-column-sorters"
                      >
                        <span
                          class="ant-table-column-title"
                        >
                          Requests Per Second
                        </span>
                        <span
                          class="ant-table-column-sorter ant-table-column-sorter-full"
                        >
                          <span
                            aria-hidden="true"
                            class="ant-table-column-sorter-inner"
                          >
                            <span
                              aria-label="caret-up"
                              class="anticon anticon-caret-up ant-table-column-sorter-up"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="caret-up"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                />
                              </svg>
                            </span>
                            <span
                              aria-label="caret-down"
                              class="anticon anticon-caret-down ant-table-column-sorter-down"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="caret-down"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                />
                              </svg>
                            </span>
                          </span>
                        </span>
                      </div>
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="ant-table-tbody"
                >
                  <tr
                    class="ant-table-row ant-table-row-level-0"
                    data-row-key="service fetch failed"
                  >
                    <td
                      class="ant-table-cell"
                    >
                      <a
                        href="/application/service fetch failed"
                        style="text-transform: capitalize;"
                      >
                        <strong>
                          service fetch failed
                        </strong>
                      </a>
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      NaN
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      NaN
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      NaN
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ServicesTable > when: the user goes to the services table and there are services > then: we render a table with the services 1`] = `
<div>
  <div
    class="ant-table-wrapper css-dev-only-do-not-override-7ny38l"
  >
    <div
      class="ant-spin-nested-loading css-dev-only-do-not-override-7ny38l"
    >
      <div
        class="ant-spin-container"
      >
        <div
          class="ant-table css-dev-only-do-not-override-7ny38l"
        >
          <div
            class="ant-table-container"
          >
            <div
              class="ant-table-content"
            >
              <table
                style="table-layout: auto;"
              >
                <colgroup />
                <thead
                  class="ant-table-thead"
                >
                  <tr>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Service
                    </th>
                    <th
                      aria-label="P99 latency (in ms)"
                      class="ant-table-cell ant-table-column-has-sorters"
                      scope="col"
                      tabindex="0"
                    >
                      <div
                        class="ant-table-column-sorters"
                      >
                        <span
                          class="ant-table-column-title"
                        >
                          P99 latency (in ms)
                        </span>
                        <span
                          class="ant-table-column-sorter ant-table-column-sorter-full"
                        >
                          <span
                            aria-hidden="true"
                            class="ant-table-column-sorter-inner"
                          >
                            <span
                              aria-label="caret-up"
                              class="anticon anticon-caret-up ant-table-column-sorter-up"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="caret-up"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                />
                              </svg>
                            </span>
                            <span
                              aria-label="caret-down"
                              class="anticon anticon-caret-down ant-table-column-sorter-down"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="caret-down"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                />
                              </svg>
                            </span>
                          </span>
                        </span>
                      </div>
                    </th>
                    <th
                      aria-label="Error Rate (in %)"
                      class="ant-table-cell ant-table-column-has-sorters"
                      scope="col"
                      tabindex="0"
                    >
                      <div
                        class="ant-table-column-sorters"
                      >
                        <span
                          class="ant-table-column-title"
                        >
                          Error Rate (in %)
                        </span>
                        <span
                          class="ant-table-column-sorter ant-table-column-sorter-full"
                        >
                          <span
                            aria-hidden="true"
                            class="ant-table-column-sorter-inner"
                          >
                            <span
                              aria-label="caret-up"
                              class="anticon anticon-caret-up ant-table-column-sorter-up"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="caret-up"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                />
                              </svg>
                            </span>
                            <span
                              aria-label="caret-down"
                              class="anticon anticon-caret-down ant-table-column-sorter-down"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="caret-down"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                />
                              </svg>
                            </span>
                          </span>
                        </span>
                      </div>
                    </th>
                    <th
                      aria-label="Requests Per Second"
                      class="ant-table-cell ant-table-column-has-sorters"
                      scope="col"
                      tabindex="0"
                    >
                      <div
                        class="ant-table-column-sorters"
                      >
                        <span
                          class="ant-table-column-title"
                        >
                          Requests Per Second
                        </span>
                        <span
                          class="ant-table-column-sorter ant-table-column-sorter-full"
                        >
                          <span
                            aria-hidden="true"
                            class="ant-table-column-sorter-inner"
                          >
                            <span
                              aria-label="caret-up"
                              class="anticon anticon-caret-up ant-table-column-sorter-up"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="caret-up"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                />
                              </svg>
                            </span>
                            <span
                              aria-label="caret-down"
                              class="anticon anticon-caret-down ant-table-column-sorter-down"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="caret-down"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                />
                              </svg>
                            </span>
                          </span>
                        </span>
                      </div>
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="ant-table-tbody"
                >
                  <tr
                    class="ant-table-row ant-table-row-level-0"
                    data-row-key="testService1"
                  >
                    <td
                      class="ant-table-cell"
                    >
                      <a
                        href="/application/testService1"
                        style="text-transform: capitalize;"
                      >
                        <strong>
                          testService1
                        </strong>
                      </a>
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      4000.00
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      0.00
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      0.50
                    </td>
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0"
                    data-row-key="testService2"
                  >
                    <td
                      class="ant-table-cell"
                    >
                      <a
                        href="/application/testService2"
                        style="text-transform: capitalize;"
                      >
                        <strong>
                          testService2
                        </strong>
                      </a>
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      2000.00
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      0.01
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      0.75
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
